Adds the possibility of using innerJoin after contain

DiogoRMax opened the pull request:
Example of use:

$Users->find()->contain(['Contacts'])->innerJoin(['CT' => 'contact_types'], ['CT.id = Contacts.contact_type_id']);

DiogoRMax committed:
Adds the possibility of using innerJoin after contain

Example of use:

$Users->find()->contain(['Contacts'])->innerJoin(['CT' => 'contact_types'], ['CT.id = Contacts.contact_type_id']);

inoas commented:
Most important would be unit tests that proof that this did not work before and that proof that it works afterwards I suppose as changes to the ORM can break real world apps quite easily.

DiogoRMax commented:
Excuses I do not know how to do these unit tests.

However the sql generated when using the current ORM is like:

> SELECT * 
> FROM users Users
> 	LEFT JOIN contact_types CT ON (
> 		CT.id = Contacts.contact_type_id
> 	)
> 	LEFT JOIN contacts Contacts ON (
> 		Contacts.user_id = (Users.id)
> 	)

What causes error

With the change it puts the innerJoin afterwards like this:

> SELECT * 
> FROM users Users
> 	LEFT JOIN contacts Contacts ON (
> 		Contacts.user_id = (Users.id)
> 	)
> 	LEFT JOIN contact_types CT ON (
> 		CT.id = Contacts.contact_type_id
> 	)


If you can help me with the process, I'm grateful ....

markstory commented:
Your changes [cause existing tests to fail](https://travis-ci.org/cakephp/cakephp/jobs/317401087) so there is likely a regression against existing behavior being introduced.

markstory commented on src/ORM/Query.php:
```diff
@@ -1059,12 +1059,20 @@ protected function _transformQuery()
+
+        $i = count($this->_parts['join']);
+        foreach ($joins as $a => $j) {
+            $this->_parts['join'][is_string($a) ? $a : $i++] = $j;
```
Couldn't this result in joins created by the eagerloader being overwritten?

DiogoRMax replied:
The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

DiogoRMax replied:
The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

DiogoRMax commented:
The problem is in the order in which JOINS are created.

The eagerloader is processed last.
Indepedent order to call the method.

DiogoRMax committed:
Revert "Adds the possibility of using innerJoin after contain"

This reverts commit 5ce89b2257d1b5ed87459c6fe672175096168325.

DiogoRMax commented:
Is there no way that the call order can be respected?

DiogoRMax commented:
I believe adding the respect to the invocation order of the method makes the ORM much more powerful and versatile.

But I also agree that it may end up breaking some apps.

The alternative would be to create a smart computer, which takes into account the dependency of JOINS.

If this change can be accepted, I can create the code.

inoas commented:
Are you certain the order of non-nested LEFT-JOINS matters as shown in your examples? As far as I know it does not.

lorenzo commented:
Have you tried using `matching()` or `innerJoinWith()`? I believe either of those will solve your problem

DiogoRMax committed:
Adding an intelligent ordering system to the query compiler

Example of use:

$Users->find()->contain(['Contacts'])->innerJoin(['CT' => 'contact_types'], ['CT.id = Contacts.contact_type_id'])->enableSmartJoin();

DiogoRMax committed:
Lint changes

DiogoRMax committed:
Lint Update

DiogoRMax commented:
Sorry for the inconvenience, I'm new to this business.

DiogoRMax commented:
How can errors occur in the tests, and the code added, should not change the flow of the program in any way?

markstory commented on src/Database/Query.php:
```diff
@@ -1926,6 +1934,31 @@ public function isBufferedResultsEnabled()
+    }
+
+    /**
+     * Returns the current hydration mode.
```
Documentation needs to be updated.

markstory commented on src/Database/QueryCompiler.php:
```diff
@@ -250,6 +255,71 @@ protected function _buildJoinPart($parts, $query, $generator)
+            $part .= $joins[$alias];
+        }
+
+        return $part;
```
This is a significant amount of code without any tests. Adding new modes increases complexity quite a bit as this mode can be combined with every other possible state. Is there a way to accomplish what you are trying to do without additional modes?

DiogoRMax replied:
I can not imagine any solution that will not break applications already deployed in the real world.

My point is:

$query->innerJoin(A...)->innerJoin(B...)
does not generate the same SQL as:
$query->innerJoin(B...)->innerJoin(A...)

But ...

$query->contain(A...)->innerJoin(B...)
$query->innerJoin(B...)->contain(A...)
Generates the same SQL, always leaving the JOIN part of the 'contain' at the end.

DiogoRMax replied:
Has anyone ever had this difficulty?

The only way I can see not to break the implementations already made, is to create an intelligent ordering system for the JOIN part.

DiogoRMax replied:
Or maybe create some parameter that holds the order of the containsJOINs

markstory replied:
Have you tried replacing `innerJoin` with `innerJoinWith` as @lorenzo suggested earlier? `innerJoinWith` uses the eagerloader so it should preserve join ordering.

> Has anyone ever had this difficulty?

Yes a very similar issue was reported in #10746 but there are a few solutions to that problem.

DiogoRMax commented:
Thank you...
But my project does not understand the use of innerJoinWith, and thinking about it is a particularity that does not have to be included in Cakephp.

I have solved my problems with other solutions, thanks for the attention ...
